### DISCUSSION TITLE
Should we add swayr? Which features?

### DISCUSSION THREAD

**Discussion by boredland** [Manjaro-Sway/manjaro-sway#184] [Ideas]
https://sr.ht/~tsdh/swayr/ is pretty amazing.

- switch-window
- quit-window

seem super useful. other may be worth a look too. wdyt?

**Comment by Zeioth** [comment]
I personaly use those two features, and find them useful. Probably most users using manjaro-sway are gonna look for something like swayr at some point.

On the other side, every dependency we add, make it the OS a bit harder to maintain too. Freezing a swayr version shouldn't be super risky, as neither sway or swayr introduce breaking changes too often.

**Comment by AdriandMartin** [comment]
It would be great if a window switcher was added by default to Manjaro Sway.

I've been using stacking window managers for a long time, and all of them had an alt-tab implementation for window switching. At the moment I'm using *workspace back_and_forth* for it, because Sway only lets switching between the current and the previous used workspace, but [no between windows](https://github.com/swaywm/sway/issues/3974) neither [having a builtin record of the most recently used windows](https://github.com/swaywm/sway/issues/4103), so the unique alternative is implementing it yourself or using an existing implementation (and swayr seems to be the best).

A window switcher is the only functionality I miss since I switched to tiling window managers.

**Reply by boredland** [reply]
great to hear! have to tried working with swayr? I'd be happy to add it to our repos (if I didn't already, cant recall). alt+tab feels like a natural bind to me too. What do others think?

**Reply by AdriandMartin** [reply]
At the moment swayr is only in AUR, so it would be great adding it to the manjaro-sway repo

**Reply by boredland** [reply]
added swayr to our repo.

**Reply by AdriandMartin** [reply]
Great, thank you! Is there any intention to add it to the default configuration? I'll be trying it these days, so I could open a PR proposing some keybindings

**Reply by boredland** [reply]
Sure, I'd be very happy if you could do that!

**Comment by AdriandMartin** [comment]
I'll try to summarize here my experience with *swayr* these days

## Configuration

Yesterday I openned a [PR](https://github.com/Manjaro-Sway/desktop-settings/pull/110) in which I propose a *config.toml* file which consist of a copy of the file generated automatically by *swayrd*, but in which `rofi` is used instead of `wofi`, and makes the menu look as follows:
![2022-06-01T16:59:31,126596084+02:00](https://user-images.githubusercontent.com/22200464/171480561-5f3595f7-1d3c-4dd9-992b-c5b56ad92569.png)

If this file doesn't exists in the path *~/.config/swayrd*, the tool's daemon will generate one in which `wofi` is used as the menu, so it doesn't work. That's why I think this file must be provided in the *skel* directory as part of the distro (I give more details about it in the commentary of the PR)

## Usage

Although *swayr* is very complete, and implements more functionalities than a simple window switcher, I've tried specially the commands related to that affair. The tool allows using two kind of switcher:
- [Menu switchers](https://git.sr.ht/~tsdh/swayr#menu-switchers), which open a menu using the command specified (in the proposed config file it's `rofi`) to generate a list with all the openned windows in LRU order (see the picture above, in which I've used `swayr switch-window`)
  
  The problem I found with this kind of switcher is that, once the menu is opened (because the key binding has been pressed), you must use the arrows or the tab key to navigate through the listed windows, because executing again the menu switcher (pressing the key binding again) causes another instance of the menu opens when the first is closed. It's really annoying having to use it that way, moving with the arrows and pressing enter to make the choice
- [Non-menu switchers](https://git.sr.ht/~tsdh/swayr#non-menu-switchers), which switch between the openned windows without showing any menu. The cyclic commands allow the same functionality than the menu switcher shown, but without generating a menu; for doing it, *swayrd* creates a list with the opened windows in LRU order when `swayr next-window` or `swayr prev-window` is executed for the first time, and stores it with its iterator so that the next execution of any of those commands be coherent with both the order and the position inside the list

  The problem I found with the cyclic commands is the need of executing `swayr nop` to break the sequence; when executing `swayr nop` you order the daemon to discard the currently iterated list, so that the next time a cyclic command is executed, it will be generated a new list in which the windows will be ordered in LRU according to the access time at that moment. Nevertheless, the project is stuck because their idea was binding the `swayr nop` execution to the *Alt* key releasing (assuming that both `swayr next-window` and `swayr prev-window` have been binded to *Alt+Tab* and *Alt+Shift+Tab*, respectively), so while iterating the windows list, the LRU order of the first execution is maintained, and when releasing the *Alt* key (it means, when the expected window is reached) the sequence breaks, making that in the next execution of a cyclic command the list has to be recalculated. The problem is that they're pending of [a Sway's issue related to key releasing events](https://github.com/swaywm/sway/issues/6456), so this functionality is not directly implementable at the moment

I'll tell you if I discover something else, but because of the reasons exposed, I think it's to soon to add any key bindings to the default Sway's configuration

**Reply by davidthewatson** [reply]
I'd vote for including swayr as you describe with its implicit limitations based on the ```nop``` problem and so forth. 

I prefer having the heuristic or semantic switcher as an additional alternative for those of us habituated into Cmd+p or Ctrl+p or $mod+p behavior that you've seen from the litany of text editors implementing it: sublime, atom, vscode, etc (perhaps others).

The other way I see of going beyond the current switcher behavior at the moment with a perhaps less complete implementation is to add sway-overfocus:

https://github.com/korreman/sway-overfocus

I've tested this and it works as described.

A third possibility is that implemented by garuda sway since I've been using it, which is to:

```
.config/sway/config.d/default:    bindsym $mod+p exec ~/.config/wofi/windows.py
```

and

```
#!/bin/python3
from argparse import ArgumentParser
import subprocess
import json

enter="\n"

# Returns a list of all json window objects
def get_windows():

    command="swaymsg -t get_tree"

    active_outputs = []

    process = subprocess.Popen(command,shell=True,stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    data = json.loads(process.communicate()[0])

    # Select outputs that are active
    windows = []
    for output in data['nodes']:

        # The scratchpad (under __i3) is not supported
        if output.get('name') != '__i3' and output.get('type') == 'output':
                workspaces = output.get('nodes')
                for ws in workspaces:
                    if ws.get('type') == 'workspace':
                        windows += extract_nodes_iterative(ws)
    return windows

# Extracts all windows from a sway workspace json object
def extract_nodes_iterative(workspace):
    all_nodes = []

    floating_nodes = workspace.get('floating_nodes')

    for floating_node in floating_nodes:
        all_nodes.append(floating_node)

    nodes = workspace.get('nodes')

    for node in nodes:

        # Leaf node
        if len(node.get('nodes')) == 0:
            all_nodes.append(node)
        # Nested node, handled iterative
        else:
            for inner_node in node.get('nodes'):
                nodes.append(inner_node)

    return all_nodes

# Returns an array of all windows
def parse_windows(windows):
    parsed_windows = []
    for window in windows:
        parsed_windows.append(window.get('name'))
    return parsed_windows

# Returns a newline seperated UFT-8 encoded string of all windows for wofi
def build_wofi_string(windows):
    return enter.join(windows).encode("UTF-8")

# Executes wofi with the given input string
def show_wofi(windows):

    command="wofi -c ~/.config/wofi/config -s ~/.config/wofi/style.css -p \"Windows: \" -d -i --hide-scroll"

    process = subprocess.Popen(command,shell=True,stdin=subprocess.PIPE,stdout=subprocess.PIPE)
    return process.communicate(input=windows)[0]

# Returns the sway window id of the window that was selected by the user inside wofi
def parse_id(windows, parsed_windows, selected):
    selected = (selected.decode("UTF-8"))[:-1] # Remove new line character
    window_index = int(parsed_windows.index(selected)) # Get index of selected window in the parsed window array
    return str(windows[window_index].get('id')) # Get sway window id based on the index

# Switches the focus to the given id
def switch_window(id):
    command="swaymsg [con_id={}] focus".format(id)

    process = subprocess.Popen(command,shell=True,stdout=subprocess.PIPE)
    process.communicate()[0]

# Entry point
if __name__ == "__main__":

    parser = ArgumentParser(description="Wofi based window switcher")

    windows = get_windows()

    parsed_windows = parse_windows(windows)

    wofi_string = build_wofi_string(parsed_windows)

    selected = show_wofi(wofi_string)

    selected_id = parse_id(windows, parsed_windows, selected)

    switch_window(selected_id)
```

I've also used the $mod+p switcher listed here on garuda sway and it does provide the quick heuristic style I prefer similar to the others above.

I'm not biased to any particular implementation. I'd probably default to whatever is simplest with least side effects from an architectural standpont.

Cheers,
Dave

**Comment by AdriandMartin** [comment]
In its last versions, `swayr` has included the "auto nop" feature; if [configured](https://sr.ht/~tsdh/swayr/#the-misc-section), it allows executing a *nop* command automatically when the specified number of milliseconds has elapsed since the last `swayr` command

Although it could seem to partially solve the problem with the cyclic commands, it doesn't work as expected, or at least that's my thought after having tested it for a few weeks. Maybe the time values I used are too short or too long, but even when I use the project's configuration (3 seconds as *auto_nop_delay* value), it doesn't do the trick for me

Relating to the @davidthewatson commentary, after the trial mentioned I've restored my keyboard shortcuts for using a swayr's menu switcher, concretely `swayr switch-window`; it makes exactly what the Garuda's script but sorting the windows in LRU order, so if you're used to that implementation you might find this command really useful. By the other hand, if that's the desired behaviour, I'll agree in adding it to the default configuration, because `swayr` seems to be the more complete and customizable option at the moment, and doing it would suppose including a functionality which other distros using Sway have already incorporated (in reference to Garuda)

**Comment by boredland** [comment]
has been added, but no too extensively.
